app/scripts/migrations/052.js

```diff
@@ -0,0 +1,163 @@
+import { cloneDeep, uniqBy } from 'lodash';
+import {
+  GOERLI,
+  GOERLI_CHAIN_ID,
+  KOVAN,
+  KOVAN_CHAIN_ID,
+  MAINNET,
+  MAINNET_CHAIN_ID,
+  NETWORK_TYPE_RPC,
+  RINKEBY,
+  RINKEBY_CHAIN_ID,
+  ROPSTEN,
+  ROPSTEN_CHAIN_ID,
+} from '../../../shared/constants/network';
+
+const version = 52;
+
+/**
+ * Migrate tokens in Preferences to be keyed by chainId instead of
+ * providerType. To prevent breaking user's MetaMask and selected
+ * tokens, this migration copies the RPC entry into *every* custom RPC
+ * chainId.
+ */
+export default {
+  version,
+  async migrate(originalVersionedData) {
+    const versionedData = cloneDeep(originalVersionedData);
+    versionedData.meta.version = version;
+    const state = versionedData.data;
+    versionedData.data = transformState(state);
+    return versionedData;
+  },
+};
+
+function transformState(state = {}) {
+  if (state.PreferencesController) {
+    const {
+      accountTokens,
+      accountHiddenTokens,
+      frequentRpcListDetail,
+    } = state.PreferencesController;
+
+    const newAccountTokens = {};
+    const newAccountHiddenTokens = {};
+
+    if (accountTokens && Object.keys(accountTokens).length > 0) {
+      for (const account of Object.keys(accountTokens)) {
+        newAccountTokens[account] = {};
+        for (const providerType of Object.keys(accountTokens[account])) {
+          switch (providerType) {
+            case MAINNET:
+              newAccountTokens[account][MAINNET_CHAIN_ID] =
+                accountTokens[account][MAINNET];
+              break;
+            case ROPSTEN:
+              newAccountTokens[account][ROPSTEN_CHAIN_ID] =
+                accountTokens[account][ROPSTEN];
+              break;
+            case RINKEBY:
+              newAccountTokens[account][RINKEBY_CHAIN_ID] =
+                accountTokens[account][RINKEBY];
+              break;
+            case GOERLI:
+              newAccountTokens[account][GOERLI_CHAIN_ID] =
+                accountTokens[account][GOERLI];
+              break;
+            case KOVAN:
+              newAccountTokens[account][KOVAN_CHAIN_ID] =
+                accountTokens[account][KOVAN];
+              break;
+            default:
+              break;
+          }
+        }
+        if (accountTokens[account][NETWORK_TYPE_RPC]) {
```

Comment:
Any reason to put this in its own `if` block instead of add it as a `case` to the above `switch`?

Reply:
Oh, because you need the above for loop to complete before the below logic can properly migrate account tokens for custom rpcs that share a chain id with one of the default networks